### PR TITLE
MPIDIG refactoring

### PR DESCRIPTION
**Pull request by gcongiu**
## Pull Request Description

The generic layer is more than just active messages. This PR refactors the current structure to prepare the addition of new generic modules.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by gcongiu** [comment]
test:jenkins/ch4/most

**Comment by gcongiu** [comment]
@hzhou this PR follows along our discussion on the generic layer abstraction and the AM code separation. I would also try to move the rest of the `ch4r` code currently in `ch4/src` into `ch4/generic/am`. Thoughts?

**Comment by hzhou** [comment]
While it is the direction I agree with, currently it will cause big conflicts with a few of my open PR. Maybe hold this for a bit?

**Comment by wesbland** [comment]
If we're going to do...sigh...more refactoring, can you please share any sort of formula you've used to make these changes so we can try to keep up with non-public branches? For instance, if this patch is *always* updating `MPIDIG_` to `MPIDIG`, that would be helpful to know so we can try to recreate this internally.

**Comment by hzhou** [comment]
> If we're going to do...sigh...more refactoring, can you please share any sort of formula you've used to make these changes so we can try to keep up with non-public branches? For instance, if this patch is _always_ updating `MPIDIG_` to `MPIDIG`, that would be helpful to know so we can try to recreate this internally.

We are waiting for you to ask :) 

The formula here is to move nearly all current `MPIDIG_` into `MPIDIG_am_` since they are all active message stuff. We are doing this because we are adding another "generic fallback" component, in this case, generic gpu stage buffer fallback.

I almost can see your "sigh" :), but trust us that we are doing this only because we are convinced it is beneficial for the "medium"-term development and certainly for the long-term maintenance. Join the discussion early and challenge/convince us the otherwise :)

**Comment by wesbland** [comment]
You're probably right that this is the right thing to do if we're going to be putting GPU code in the same place. I'm just imagining the merge conflicts now. As long as you can keep this PR as pretty much only changing file names and function/datatype prefixes, it makes merging things simpler.

**Comment by gcongiu** [comment]
@wesbland sure we can keep this simple and leave all the restructuring of active message code in `ch4/src` to a later PR, if eventually needed.

**Comment by wesbland** [comment]
> @wesbland sure we can keep this simple and leave all the restructuring of active message code in `ch4/src` to a later PR, if eventually needed.

Thanks. That should help us merge things a bit.

**Comment by gcongiu** [comment]
test:jenkins/ch4/most

**Comment by gcongiu** [comment]
test:mpich/warnings

**Comment by gcongiu** [comment]
test:mpich/warnings

**Comment by gcongiu** [comment]
@yfguo refer to this PR https://github.com/pmodels/mpich/pull/4303 to see I we can extend the generic layer after this refactoring.

File: src/mpid/ch4/generic/Makefile.mk [hunk 1]
```diff
@@ -1,21 +1,7 @@
-## -*- Mode: Makefile; -*-
```
Comment (yfguo) [inline src/mpid/ch4/generic/Makefile.mk]: Please keep the modeline for emacs

**Comment by gcongiu** [comment]
@yfguo is the namespace change ok? Or should I leave it for another PR?

**Comment by yfguo** [comment]
I think it would be better to put off the name changes to a later day. We could settle with `MPIDIG_` for the existing AM and add `MPIDIG_gpu` for GPU related code. Then do a one-time namespace change to fix the namespaces.

**Comment by gcongiu** [comment]
test:jenkins/ch4/most
test:mpich/warnings

**Comment by gcongiu** [comment]
test:jenkins/ch4/most
test:mpich/warnings

File: src/mpid/ch4/generic/am/mpidig_am_globals.c [hunk 2]
```diff
@@ -10,11 +10,10 @@
  */
 
 #include "mpidimpl.h"
-#include "mpidig.h"
 #include "ch4_impl.h"
 #include "mpidch4r.h"
 
-MPIDIG_global_t MPIDIG_global = { {0}
+MPIDIG_am_global_t MPIDIG_am_global = { {0}
```
Comment (gcongiu) [inline src/mpid/ch4/generic/am/mpidig_am_globals.c]: I realized `MPIDIG_global_t` is also AM specific. Eventually there should be an `MPIDIG_global_t`  in `ch4/generic` folder that includes `MPIDIG_am_global`, `MPIDIG_gpu_global`, ...

**Comment by gcongiu** [comment]
test:jenkins/ch4/most
test:mpich/warnings

**Comment by gcongiu** [comment]
test:jenkins/ch4/most
test:mpich/warnings

**Comment by gcongiu** [comment]
@yfguo I have removed namespace changes exception made for `MPIDIG_am_init/finalize` as I need to disambiguate with `MPIDIG_init`, which will eventually call both `MPIDIG_am_init` and `MPIDIG_gpu_init` during initialization. I also kept filename changes for `am` code.

**Pull request merged**